swift/README.md

```diff
@@ -1,372 +1,1100 @@
 
-// good
-if condition {
-    return
+Use implicitly unwrapped types declared with `!` only for instance variables that you know will be initialized later before use, such as subviews that will be set up in `viewDidLoad`. Other cases you shoudl use `!` are:
+
+- You just assigned to an object where you know initialization can not fail:
+
+```swift
+let numberSeven = Int("7")! //Don't do this.
+```
+
+In this case, wrap this into a static function and write a test for it:
+
+```swift
+extension Int {
+    static func seven() -> Int {
+        return Int("7")!
+    }
 }
+```
 
-// bad
-guard let x = x else { return }
+```swift
+import XCTest
 
-// good
-guard let x = x else {
-    return
+class IntegerTests: XCTestCase {
+    func testSeven() {
+        XCTAssertEqual(7, Int.seven())
+    }
 }
 ```
 
-Objective-C interop
-===================
+Although the test might seem redundant, because `seven()` returns a non-optional, this ensures that even if in the future someone changes the implementation of `seven()`, it won't get past CI if the explicit unwrapping would fail.
 
-Do not use `@objc` or inherit from `NSObject` unless absolutely necessary.
+- You are creating an object during initialization but need to pass `self` to that object's initializer:
 
-Singletons
-==========
+```swift
+class Component: ParentComponent {
+    var controller: CustomViewController! // We want to use it as a non-optional but have to initialize after super.init()
 
-Avoid this design pattern.
+    override init() {
+        // controller = CustomViewController(component: self) <-- compiler error
+        super.init()
+        controller = CustomViewController(component: self)
+    }
+}
+```
 
-`let` vs `var`
-==============
+When accessing an optional value, use optional chaining if the value is only accessed once or if there are many optionals in the chain:
 
-Use `let` unless the variable will be mutated.
+```swift
+self.textContainer?.textLabel?.setNeedsDisplay()
+```
 
-delegates
-=========
+Use optional binding when it's more convenient to unwrap once and perform multiple operations:
 
-Use the `weak` modifier for `delegate` properties.  
+```swift
+if let textContainer = self.textContainer {
+  // do many things with textContainer
+}
+```
 
-Name the delegate property `delegate` unless your class supports multiple different delegates.  
+When naming optional variables and properties, avoid naming them like `optionalString` or `maybeView` since their optional-ness is already in the type declaration.
 
-Protocols (aka Interfaces)
-==========================
+For optional binding, shadow the original name when appropriate rather than using names like `unwrappedView` or `actualLabel`.
 
-Protocols are awesome. Use them to restrict the API surface area of an object
-being passed around.
+Use `as` for type coercion if possible (this is enforced statically). Otherwise use `as?`.
```

Comment:
Umm, I think the difference is between guaranteed conversion (`as`, upcasting, verified by the compiler that it will succeed) and forced conversion (`as?`, `as!`, downcasting, can't be verified by the compiler). So this rule is intended to be "use `as` when possible (upcasting), if you can't use `as`, use `as?` (for downcasting), but don't use `as!`. Maybe I can make it a little clearer?